File: shared/src/commonMain/kotlin/energy/ftw/carrier/Carrier.kt
package energy.ftw.carrier

enum class CarrierKind { Relay, Cache, None }

sealed class CarrierStatus {
    data object Connecting : CarrierStatus()
    data class Open(val sinceMs: Long) : CarrierStatus()
    data class Closed(val reason: String, val retryable: Boolean) : CarrierStatus()
}

interface Carrier {
    val kind: CarrierKind
    val status: CarrierStatus
    fun send(frame: ByteArray)
    fun onFrame(handler: (ByteArray) -> Unit): () -> Unit
    fun onStatus(handler: (CarrierStatus) -> Unit): () -> Unit
    fun wake() {}
    fun close(reason: String = "closed by client")
}

open class CarrierBase : Carrier {
    override val kind: CarrierKind = CarrierKind.None
    override var status: CarrierStatus = CarrierStatus.Connecting
        protected set

    private val frames = linkedSetOf<(ByteArray) -> Unit>()
    private val statuses = linkedSetOf<(CarrierStatus) -> Unit>()

    override fun send(frame: ByteArray) {}

    override fun onFrame(handler: (ByteArray) -> Unit): () -> Unit {
        frames.add(handler)
        return { frames.remove(handler) }
    }

    override fun onStatus(handler: (CarrierStatus) -> Unit): () -> Unit {
        statuses.add(handler)
        return { statuses.remove(handler) }
    }

    override fun close(reason: String) {
        emitStatus(CarrierStatus.Closed(reason, retryable = false))
        frames.clear()
        statuses.clear()
    }

    protected fun emitFrame(frame: ByteArray) {
        for (h in frames.toList()) {
            try {
                h(frame)
            } catch (_: Throwable) {
            }
        }
    }

    protected fun emitStatus(s: CarrierStatus) {
        status = s
        for (h in statuses.toList()) {
            try {
                h(s)
            } catch (_: Throwable) {
            }
        }
    }
}

interface RawSocket {
    fun sendBinary(bytes: ByteArray)
    fun close()
}

interface SocketListener {
    fun onOpen()
    fun onBinary(bytes: ByteArray)
    fun onText(text: String)
    fun onClose(code: Int, reason: String)
}

fun interface SocketFactory {
    fun open(url: String, listener: SocketListener): RawSocket
}

class LoopbackCarrier : CarrierBase() {
    override val kind = CarrierKind.Relay

    fun peerSend(frame: ByteArray) = emitFrame(frame)

    fun markOpen() = emitStatus(CarrierStatus.Open(nowMs()))

    override fun send(frame: ByteArray) {}
}

File: appleApp/FTWKit/Sources/FTWKit/Carrier/Carrier.swift
import Foundation

/// How frames are reaching us. Orthogonal to `SourceState`, never merged
/// with it: "connected, but the inverter went quiet 40 seconds ago" needs
/// both.
public enum CarrierKind: String, Sendable, Codable {
    case webrtc, relay, cache, none
}

public enum CarrierStatus: Equatable, Sendable {
    case connecting
    case open(sinceMs: Double)
    case closed(reason: String, retryable: Bool)

    public var isOpen: Bool {
        if case .open = self { return true }
        return false
    }

    var phaseName: String {
        switch self {
        case .connecting: return "connecting"
        case .open: return "open"
        case .closed: return "closed"
        }
    }
}

/// Moves opaque frames and nothing else. It cannot read a frame and has no
/// opinion about the protocol, so a LAN path can be added later as one more
/// implementation without a line changing above this.
@MainActor
public protocol Carrier: AnyObject {
    var kind: CarrierKind { get }
    var status: CarrierStatus { get }
    func send(_ frame: Bytes)
    /// Replaces the handlers. One owner at a time: the carrier stack is
    /// built as a chain, each layer owning the one below it.
    func setHandlers(onFrame: @escaping @MainActor (Bytes) -> Void, onStatus: @escaping @MainActor (CarrierStatus) -> Void)
    /// Recheck the live path after the app or network wakes. Frames are
    /// never replayed.
    func wake()
    func close(reason: String)
}

// MARK: - WebSockets

/// The socket underneath the relay carrier, reduced to what it uses.
@MainActor
public protocol WebSocketConnection: AnyObject {
    func send(_ data: Bytes)
    func close()
}

@MainActor
public struct WebSocketEvents {
    public var onOpen: @MainActor () -> Void
    public var onText: @MainActor (String) -> Void
    public var onBinary: @MainActor (Bytes) -> Void
    /// Close code and reason. A socket that failed to open closes with 1006.
    public var onClose: @MainActor (Int, String) -> Void
}

public typealias WebSocketFactory = @MainActor (_ url: URL, _ events: WebSocketEvents) -> WebSocketConnection
